### ISSUE TITLE
truffle migrate. but it is failed!Error: Returned error: VM Exception while processing transaction: invalid opcode. how to solve it! thanks.

### ISSUE THREAD

**Issue by ouxiaobai** [trufflesuite/truffle#4058]
## Issue

Saving migration to chain.
Error: Returned error: VM Exception while processing transaction: invalid opcode

## Steps to Reproduce

Ganache CLI v6.1.8 (ganache-core: 2.2.1) starting.

## Actual Results

Compiling your contracts...
===========================
> Everything is up to date, there is nothing to compile.

Starting migrations...
======================
> Network name:    'development'
> Network id:      1621519682863
> Block gas limit: 6721975 (0x6691b7)

1_initial_migration.js
======================

   Deploying 'Migrations'
   ----------------------
   > transaction hash:    0x8f7fb36b70f9cd32bc83c6d78bc0cbe6486ad74928c4a4ed4ee36b8dd2a421f8
   > Blocks: 0            Seconds: 0
   > contract address:    0x196D74B498b502dbC1d9Fc10a39c9A8531e32C29
   > block number:        5
   > block timestamp:     1621520943
   > account:             0xde52B5B8e3A49f94885833911Cd153d058274fb6
   > balance:             99.71915058
   > gas used:            221555 (0x36173)
   > gas price:           20 gwei
   > value sent:          0 ETH
   > total cost:          0.0044311 ETH

   ⠙ Saving migration to chain.
Error: Returned error: VM Exception while processing transaction: invalid opcode
    at Migration._deploy (/usr/local/lib/node_modules/truffle/build/webpack:/packages/migrate/Migration.js:103:1)
    at process._tickCallback (internal/process/next_tick.js:68:7)
Truffle v5.3.6 (core: 5.3.6)
Node v10.19.0

 Is the reason  about the version of Ganache ?

## Environment

* Operating System: 
* Ethereum client:
* Truffle version (`v5.3.6 `):
* node version (`v10.19.0`):
* npm version (`6.14.4`): 
* Ganache(`v6.1.8`);

**Comment by haltman-at** [comment]
Versions of Solidity prior to 0.8.0 will trigger an invalid opcode on various errors.  This doesn't seem like a Truffle bug, rather it seems like something went wrong during deployment, or there's an error in the contract.

**Comment by ouxiaobai** [comment]
> Hmm, I'm not sure why this error would be caused...but perhaps it could be an old bug that was fixed. Try upgrading Ganache to the latest version to see if it solves the issue. If not I'll be happy to try and help you troubleshoot.
I updated Ganache and then deployed the contract to Ganache successfully. The reason for the failure is the version of ganache.

**Comment by ouxiaobai** [comment]
> Hmm, I'm not sure why this error would be caused...but perhaps it could be an old bug that was fixed. Try upgrading Ganache to the latest version to see if it solves the issue. If not I'll be happy to try and help you troubleshoot.

I updated Ganache and then deployed the contract to Ganache successfully. The reason for the failure is the version of ganache.
thanks.

**Comment by ouxiaobai** [comment]
> lidity prior to 0.8.0 will trigger an invalid opcode on various errors. This doesn't

**Comment by ouxiaobai** [comment]
> Versions of Solidity prior to 0.8.0 will trigger an invalid opcode on various errors. This doesn't seem like a Truffle bug, rather it seems like something went wrong during deployment, or there's an error in the contract.

There should be no mistakes in the contract. I use the contract of the metacoin project on the official website.
thanks.

**Comment by haltman-at** [comment]
Hm, yes, I suppose this error could occur if you used an old version of Ganache that didn't support the hardfork the contract was compiled for.

**Comment by fainashalts** [comment]
Closing because it sounds like this is resolved. Let us know if that's not the case, thanks!

**Comment by vochithanh** [comment]
My case: using hardfork in Ganache config: Byzantium -> error, switch to hardfork: Petersburg -> ok.

**Comment by fainashalts** [comment]
@vochithanh thanks for bringing this to our attention. Since this issue is closed and your issue seems like it may be specifically related to Ganache functionality, could you write up a new issue for what you are seeing [here] (https://github.com/trufflesuite/ganache/issues)? Thanks so much!

**Comment by Arnab28122000** [comment]
try to restart ganache, it worked for me

**Comment by Starbet** [comment]
> My case: using hardfork in Ganache config: Byzantium -> error, switch to hardfork: Petersburg -> ok.

this worked for me
